From e4158a43728d7e6e5b0bcf4eefff269b32bf7803 Mon Sep 17 00:00:00 2001
From: Peter Chen <peter.chen@nxp.com>
Date: Thu, 21 Dec 2017 09:06:40 +0800
Subject: [PATCH 3129/5242] MLK-17261-2 usb: chipidea: imx: do not call hsic
 callback for non-hsic controller

commit  5e1b455775e06aae11cb608b0d526c57bf2593d1 from
https://source.codeaurora.org/external/imx/linux-imx.git

With this judgement, the non-hsic controller will access wrong registers,
and below error message will be showed:
"usbmisc_imx 2184800.usbmisc: index is error for usbmisc"

Fixes: 113be1516160 ("MLK-16715-6 usb: chipidea: imx:
	add HSIC support for controllers from imx7d")
Reported-by: Peng Fan <peng.fan@nxp.com>
Acked-by: Jun Li <jun.li@nxp.com>
Signed-off-by: Peter Chen <peter.chen@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/usb/chipidea/ci_hdrc_imx.c |    1 +
 drivers/usb/chipidea/usbmisc_imx.c |    4 ++--
 2 files changed, 3 insertions(+), 2 deletions(-)

diff --git a/drivers/usb/chipidea/ci_hdrc_imx.c b/drivers/usb/chipidea/ci_hdrc_imx.c
index f122078..03a88a5 100644
--- a/drivers/usb/chipidea/ci_hdrc_imx.c
+++ b/drivers/usb/chipidea/ci_hdrc_imx.c
@@ -1,6 +1,7 @@
 // SPDX-License-Identifier: GPL-2.0+
 /*
  * Copyright 2012-2016 Freescale Semiconductor, Inc.
+ * Copyright 2017 NXP
  * Copyright (C) 2012 Marek Vasut <marex@denx.de>
  * on behalf of DENX Software Engineering GmbH
  */
diff --git a/drivers/usb/chipidea/usbmisc_imx.c b/drivers/usb/chipidea/usbmisc_imx.c
index 6de412d..6d6393b 100644
--- a/drivers/usb/chipidea/usbmisc_imx.c
+++ b/drivers/usb/chipidea/usbmisc_imx.c
@@ -1204,7 +1204,7 @@ int imx_usbmisc_hsic_set_connect(struct imx_usbmisc_data *data)
 		return 0;
 
 	usbmisc = dev_get_drvdata(data->dev);
-	if (!usbmisc->ops->hsic_set_connect)
+	if (!usbmisc->ops->hsic_set_connect || !data->hsic)
 		return 0;
 	return usbmisc->ops->hsic_set_connect(data);
 }
@@ -1218,7 +1218,7 @@ int imx_usbmisc_hsic_set_clk(struct imx_usbmisc_data *data, bool on)
 		return 0;
 
 	usbmisc = dev_get_drvdata(data->dev);
-	if (!usbmisc->ops->hsic_set_clk)
+	if (!usbmisc->ops->hsic_set_clk || !data->hsic)
 		return 0;
 	return usbmisc->ops->hsic_set_clk(data, on);
 }
-- 
1.7.9.5

